<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenomicOffsets.jl Documentation · GenomicOffsets</title><meta name="title" content="GenomicOffsets.jl Documentation · GenomicOffsets"/><meta property="og:title" content="GenomicOffsets.jl Documentation · GenomicOffsets"/><meta property="twitter:title" content="GenomicOffsets.jl Documentation · GenomicOffsets"/><meta name="description" content="Documentation for GenomicOffsets."/><meta property="og:description" content="Documentation for GenomicOffsets."/><meta property="twitter:description" content="Documentation for GenomicOffsets."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicOffsets</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>GenomicOffsets.jl Documentation</a><ul class="internal"><li><a class="tocitem" href="#TL;TR"><span>TL;TR</span></a></li><li><a class="tocitem" href="#References"><span>References</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>GenomicOffsets.jl Documentation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GenomicOffsets.jl Documentation</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/currocam/GenomicOffsets.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/currocam/GenomicOffsets.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicOffsets.jl-Documentation"><a class="docs-heading-anchor" href="#GenomicOffsets.jl-Documentation">GenomicOffsets.jl Documentation</a><a id="GenomicOffsets.jl-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicOffsets.jl-Documentation" title="Permalink"></a></h1><p>Genomic offsets (GO) statistics measure current individuals/populations&#39; maladaptation in the event of a drastic climate change. These metrics are obtained from Genotype environment associations (GEA). </p><p>This Julia package provides efficient implementations of the four most popular metrics: <a href="https://doi.org/10.1111/mec.13889">RONA</a>, <a href="https://doi.org/10.1111/2041-210X.13722">RDA GO</a>, <a href="https://doi.org/10.1111/ele.12376">Gradient Forest GO</a> and <a href="https://doi.org/10.1093/molbev/msad140">Geometric GO</a>. In addition, we have implemented an F-statistical test based on the <a href="https://doi.org/10.1093/molbev/msz008">Linear Mixed Latent Factor model</a> (LFMM) to perform the identification of GEA candidates through hypothesis testing and a bootstrap approach that resamples loci, performs the GEA candidates identification and computes the desired genomic offset metric. </p><p>We have documented several &quot;recipes&quot; in both Julia and R in the <a href="https://github.com/currocam/GenomicOffsets.jl">README file of the repository</a> (so please take a look there and use this documentation as a reference for all available functions).</p><h2 id="TL;TR"><a class="docs-heading-anchor" href="#TL;TR">TL;TR</a><a id="TL;TR-1"></a><a class="docs-heading-anchor-permalink" href="#TL;TR" title="Permalink"></a></h2><p>You can compute different genomic offsets by fitting different models and computing the genomic offset.</p><pre><code class="language-julia hljs"># Julia
using GenomicOffsets
Y, X, Xpred = godataset
rona = fit(RONA, Y, X)
genomic_offset(rona, X, Xpred)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">100×1 Matrix{Float64}:
 0.04930412355509537
 0.09685110436545971
 0.031955998579532574
 0.057903936496619686
 0.06747793104523755
 0.07838283024137747
 0.034071076534702927
 0.05671776002159728
 0.11209514598766128
 0.050635672102255044
 ⋮
 0.05847056037198923
 0.044459426014584816
 0.1000681402606062
 0.076366019095105
 0.0779261159151334
 0.09861395620054535
 0.06261235895467696
 0.05449853921861244
 0.07504407351624044</code></pre><p>You can fit LFMM models to do hypothesis testing and identify GEA candidates. </p><pre><code class="language-julia hljs"># Julia
lfmm = RidgeLFMM(Y, X, 1)
pvalues = LFMM_Ftest(lfmm, Y, X; genomic_control=true)
# Bonferroni correction at 0.05
candidates = findall(pvalues .&lt; 0.05 / size(Y, 2))</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">90-element Vector{Int64}:
   52
   86
  105
  176
  211
  233
  284
  323
  412
  419
    ⋮
 4107
 4319
 4383
 4688
 4831
 4844
 4907
 4973
 5124</code></pre><p>Yo can compute bootstrapping confidence intervals for different genomic offset metrics using an approach we propose that sample <em>loci</em> and identifies GEA candidates at each iteration: </p><pre><code class="language-julia hljs">using Random, StatsBase
nboots = 100
offsets = bootstrap_with_candidates(GeometricGO, Y, X, Xpred, nboots;candidates_threshold=0.05)
@assert size(offsets) == (size(Y, 1), nboots)
confint95 = [quantile(ind, [0.025, 0.975]) for ind in eachrow(offsets)]</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">100-element Vector{Vector{Float64}}:
 [0.02303280401566295, 0.03151558990536644]
 [0.1047729314923714, 0.15591623894136308]
 [0.011123446136442448, 0.015183760680373653]
 [0.0386068546151271, 0.05458975227698637]
 [0.04292727312732361, 0.06134203909699572]
 [0.0928716599812431, 0.13115510306701367]
 [0.014603083635330173, 0.020300992040981226]
 [0.03722872857179291, 0.04973468787500883]
 [0.08909070888028804, 0.1308675409351474]
 [0.03431001066987424, 0.05007488604751634]
 ⋮
 [0.021821315064466877, 0.029341761345400032]
 [0.029222879302603963, 0.04309757900855938]
 [0.1397961718143955, 0.2056429635597111]
 [0.04246632188729315, 0.06550399094959608]
 [0.048245602602202856, 0.07065170052663027]
 [0.05532866396795552, 0.07894162577477457]
 [0.04981200290106696, 0.07009733123191236]
 [0.02786753998320747, 0.04076474730946965]
 [0.06894502320101961, 0.10013528792399119]</code></pre><h2 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.GeometricGO" href="#GenomicOffsets.GeometricGO"><code>GenomicOffsets.GeometricGO</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Geometric genomic offset. TODO: Add reference</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/Geometric_GO.jl#L1-L3">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.GradientForestGO" href="#GenomicOffsets.GradientForestGO"><code>GenomicOffsets.GradientForestGO</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Gradient genomic offset. TODO: Add reference</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GradientForest_GO.jl#L2-L4">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.LFMM" href="#GenomicOffsets.LFMM"><code>GenomicOffsets.LFMM</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">LFMM{T&lt;:Real}</code></pre><p>A struct to store the results of the Linear Fixed Effects Mixed Model (LFMM) as described by Caye et al. (2019).</p><p class="math-container">\[Y = X B^T + W = X B^T + U V^T\]</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/LFMM.jl#L35-L43">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.RDAGO" href="#GenomicOffsets.RDAGO"><code>GenomicOffsets.RDAGO</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Redundancy Analysis (RDA) genomic offset. TODO: Add reference</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RDA_GO.jl#L5-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.RONA" href="#GenomicOffsets.RONA"><code>GenomicOffsets.RONA</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Risk of non-adaptnesss (RONA) genomic offset. First proposed by Rellstab et al. (2016).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RONA.jl#L2-L4">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GenomicOffsets.LFMM, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GenomicOffsets.LFMM, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.LFMM_Ftest</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">LFMM_Ftest(model::LFMM, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; genomic_control::Bool=true, center=true) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Compute F-Statistic (F-test) for the Linear Fixed Effects Mixed Model (LFMM). TODO: Add reference &amp; description.</p><p><strong>Arguments</strong></p><ul><li><code>model</code>: A <code>LFMM{T&lt;:Real}</code> data structure (obtained from <code>RidgeLFMM</code>).</li><li><code>Y</code>: A centered genotype matrix of size NxL.</li><li><code>X</code>: A centered environmental matrix of size NxP.</li><li><code>genomic_control</code>: If true, apply genomic control to the F-statistic.</li><li><code>center</code>: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.</li></ul><p><strong>Returns</strong></p><ul><li>A vector of p-values for loci in the genotype matrix.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GEACandidates.jl#L2-L15">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GeometricGO, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GeometricGO, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.LFMM_Ftest</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">LFMM_Ftest(model::LFMM, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; genomic_control::Bool=true, center=true) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Compute F-Statistic (F-test) for the Linear Fixed Effects Mixed Model (LFMM). TODO: Add reference &amp; description.</p><p><strong>Arguments</strong></p><ul><li><code>model</code>: A <code>LFMM{T&lt;:Real}</code> data structure (obtained from <code>RidgeLFMM</code>).</li><li><code>Y</code>: A centered genotype matrix of size NxL.</li><li><code>X</code>: A centered environmental matrix of size NxP.</li><li><code>genomic_control</code>: If true, apply genomic control to the F-statistic.</li><li><code>center</code>: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.</li></ul><p><strong>Returns</strong></p><ul><li>A vector of p-values for loci in the genotype matrix.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GEACandidates.jl#L45-L58">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.RidgeLFMM-Union{Tuple{T2}, Tuple{T1}, Tuple{Matrix{T1}, Matrix{T2}, Int64}, Tuple{Matrix{T1}, Matrix{T2}, Int64, Any}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.RidgeLFMM-Union{Tuple{T2}, Tuple{T1}, Tuple{Matrix{T1}, Matrix{T2}, Int64}, Tuple{Matrix{T1}, Matrix{T2}, Int64, Any}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.RidgeLFMM</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">RidgeLFMM(Y::Matrix{T}, X::Matrix{T}, K::Int, λ::Float64) where T&lt;:Real</code></pre><p>Ridge solutions for the Linear Fixed Effects Mixed Model (LFMM) as described by Caye et al. (2019).</p><p class="math-container">\[Y = X B^T + W = X B^T + U V^T\]</p><p><strong>Arguments</strong></p><ul><li><code>Y</code>: A centered genotype matrix of size NxL.</li><li><code>X</code>: Environmental matrix of size NxP.</li><li><code>K</code>: Number of latent factors.</li><li><code>λ</code>: Regularization parameter (by 1e-5)</li><li><code>center</code>: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.</li></ul><p><strong>Returns</strong></p><ul><li>A <code>LFMM{T&lt;:Real}</code> data structure with the latent factors <code>U</code>, <code>Vt</code>, and the effect sizes <code>Bt</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/LFMM.jl#L50-L68">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.TracyWidom-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:Real" href="#GenomicOffsets.TracyWidom-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T&lt;:Real"><code>GenomicOffsets.TracyWidom</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">TracyWidom(eigenvalues::AbstractVector{T}; sort_eigenvalues::Bool=true) where T&lt;:Real</code></pre><p>Compute the Tracy-Widom statistics and p-values for a given set of eigenvalues. TODO: add reference. </p><p><strong>Arguments</strong></p><ul><li><code>eigenvalues::AbstractVector{T}</code>: A vector of eigenvalues. </li><li><code>sort_eigenvalues::Bool=true</code>: If true, sort the eigenvalues in descending order. If false, the eigenvalues are assumed to be sorted in descending order.</li></ul><p><strong>Returns</strong></p><ul><li>A tuple with the Tracy-Widom statistics and p-values. For a given threshold, it is advised not to keep all eigenvalues below the threshold, but up to the first one that is above the threshold (not included).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/LFMM.jl#L3-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.bootstrap_with_candidates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">bootstrap_with_candidates(::Type{GeometricGO}, rng::Random.AbstractRNG,Y::AbstractMatrix{T1}, X::AbstractMatrix{T2},Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Compute the Geometric genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of <code>Y</code>. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.</p><p><strong>Arguments</strong></p><ul><li><code>rng::Random.AbstractRNG</code>: Random number generator. If not provided, the global RNG is used.</li><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T2}</code>: NxP matrix with environmental data.</li><li><code>Xpred::AbstractMatrix{T2}</code>: NxP matrix with the predicted / altered environmental matrix.</li><li><code>nboot::Int=500</code>: Number of bootstrap iterations.</li><li>Optional arguments:<ul><li><code>candidates_threshold::Real=0.05</code>: Threshold for the F-test. Better to be permisive.</li><li><code>genomic_control::Bool=true</code>: Apply genomic control to the F-test.</li><li><code>tw_threshold::Real=0.001</code>: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).</li><li>λ::Real=1e-5: Regularization parameter for the LFMM.</li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/Geometric_GO.jl#L121-L140">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.bootstrap_with_candidates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">bootstrap_with_candidates(::Type{GradientForestGO}, rng::Random.AbstractRNG,Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; ntrees::Int=100, candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Compute the Gradient Forest genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of <code>Y</code>. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.</p><p><strong>Arguments</strong></p><ul><li><code>rng::Random.AbstractRNG</code>: Random number generator. If not provided, the global RNG is used.</li><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T2}</code>: NxP matrix with environmental data.</li><li><code>Xpred::AbstractMatrix{T2}</code>: NxP matrix with the predicted / altered environmental matrix.</li><li><code>nboot::Int=500</code>: Number of bootstrap iterations.</li><li>Optional arguments:<ul><li><code>ntrees::Int=100</code>: Number of trees each forest (one per allele) will have.</li><li><code>candidates_threshold::Real=0.05</code>: Threshold for the F-test. Better to be permisive.</li><li><code>genomic_control::Bool=true</code>: Apply genomic control to the F-test.</li><li><code>tw_threshold::Real=0.001</code>: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).</li><li>λ::Real=1e-5: Regularization parameter for the LFMM.</li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GradientForest_GO.jl#L51-L71">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.bootstrap_with_candidates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">bootstrap_with_candidates(::Type{RDAGO}, rng::Random.AbstractRNG, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1&lt;:Real, T2&lt;:Real}</code></pre><p>Compute the RDA genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of <code>Y</code>. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.</p><p><strong>Arguments</strong></p><ul><li><code>rng::Random.AbstractRNG</code>: Random number generator. If not provided, the global RNG is used.</li><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T2}</code>: NxP matrix with environmental data.</li><li><code>Xpred::AbstractMatrix{T2}</code>: NxP matrix with the predicted / altered environmental matrix.</li><li><code>nboot::Int=500</code>: Number of bootstrap iterations.</li><li>Optional arguments:<ul><li><code>candidates_threshold::Real=0.05</code>: Threshold for the F-test. Better to be permisive.</li><li><code>genomic_control::Bool=true</code>: Apply genomic control to the F-test.</li><li><code>tw_threshold::Real=0.001</code>: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).</li><li><code>pratio::Float64=0.99</code>: The ratio of variances preserved in the principal subspace.</li><li><code>weighted::Bool=false</code>: If true, the euclidean distance in the projected space is weighted based on the eigenvalue of each axis.</li><li>λ::Real=1e-5: Regularization parameter for the LFMM.</li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RDA_GO.jl#L64-L85">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.bootstrap_with_candidates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">bootstrap_with_candidates(::Type{RONA}, rng::Random.AbstractRNG, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1&lt;:Real, T2&lt;:Real}</code></pre><p>Compute the genomic offset for the RONA model using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of <code>Y</code>. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.</p><p><strong>Arguments</strong></p><ul><li><code>rng::Random.AbstractRNG</code>: Random number generator. If not provided, the global RNG is used.</li><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T2}</code>: NxP matrix with environmental data.</li><li><code>Xpred::AbstractMatrix{T2}</code>: NxP matrix with the predicted / altered environmental matrix.</li><li><code>nboot::Int=500</code>: Number of bootstrap iterations.</li><li>Optional arguments:<ul><li><code>candidates_threshold::Real=0.05</code>: Threshold for the F-test. Better to be permisive.</li><li><code>genomic_control::Bool=true</code>: Apply genomic control to the F-test.</li><li><code>tw_threshold::Real=0.001</code>: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).</li><li>λ::Real=1e-5: Regularization parameter for the LFMM.</li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RONA.jl#L47-L67">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}, Union{Nothing, Int64}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}, Union{Nothing, Int64}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(::Type{GeometricGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Fit the Geometric genomic offset model (that is, fit a LFMM) using the data <code>Y</code> and <code>X</code>.</p><p><strong>Arguments</strong></p><ul><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li><li><code>K</code>: Number of latent factors to use when fitting th  e LFMM model. It can be determined using the elbow rule of thumb. If nothing is provided, the Tracy-Widom statistic to determine the number of latent factors (which might be overconservative).</li><li>Optional arguments:<ul><li><code>center::Bool=true</code>: Center both the genotype and environmental data. Predicted environmental data is centered using the same mean as X. If false, data is assumed to be centered.</li><li><code>scale::Bool=true</code>: Scale the environmental data. Predicted environmental data is scaled using the same standard deviation as X. If false, data is assumed to be scaled.</li><li>tw_threshold::Real=1e-3: Threshold to determine the number of latent factors if using the Tracy-Widom statistic. By default, it uses 1e-3.</li><li>λ::Real=1e-5: Regularization parameter for the LFMM.</li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A GeometricGO model.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/Geometric_GO.jl#L12-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(::Type{GradientForestGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; ntrees::Int=100) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Fit the Gradient forest genomic offset model (that is, fit a Gradient Forest) using the data <code>Y</code> and <code>X</code>.</p><p><strong>Arguments</strong></p><ul><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li><li><code>ntrees</code>: Number of trees each forest (one per allele) will have.</li></ul><p><strong>Returns</strong></p><ul><li>A GradientForestGO model.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GradientForest_GO.jl#L9-L20">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(::Type{RDAGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Fit the RDA model (that is, apply Redundancy Analysis) using the data <code>Y</code> and <code>X</code>.</p><p><strong>Arguments</strong></p><ul><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li><li>Optional arguments:<ul><li><code>K::Int=size(Y,2)</code>: Number of main canonical axes to retain. The final number of axes is determined either by K or pratio. </li><li><code>pratio::Float64=0.99</code>: The ratio of variances preserved in the principal subspace. </li></ul></li></ul><p><strong>Returns</strong></p><ul><li>A RDAGO model.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RDA_GO.jl#L13-L26">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}" href="#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1&lt;:Real, T2&lt;:Real}"><code>GenomicOffsets.fit</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(::Type{RONA}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1&lt;:Real,T2&lt;:Real}</code></pre><p>Fit the RONA model (that is, solve the least squares problem) using the data <code>Y</code> and <code>X</code>.</p><p><strong>Arguments</strong></p><ul><li><code>Y::AbstractMatrix{T1}</code>: NxL matrix with individuals genotypes or allele frequencies.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li></ul><p><strong>Returns</strong></p><ul><li>A RONA model.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RONA.jl#L9-L19">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.genomic_offset-Union{Tuple{T2}, Tuple{T}, Tuple{GeometricGO{T}, AbstractMatrix{T}, AbstractMatrix{T}, AbstractVector{T2}}} where {T&lt;:Real, T2&lt;:Integer}" href="#GenomicOffsets.genomic_offset-Union{Tuple{T2}, Tuple{T}, Tuple{GeometricGO{T}, AbstractMatrix{T}, AbstractMatrix{T}, AbstractVector{T2}}} where {T&lt;:Real, T2&lt;:Integer}"><code>GenomicOffsets.genomic_offset</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">genomic_offset(model::GeometricGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T&lt;:Real</code></pre><p>Compute the Geometric genomic offset.</p><p><strong>Arguments</strong></p><ul><li><code>model::GeometricGO</code>: A GeometricGO model.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.</li><li><code>Xpred::Matrix{T}</code>: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.</li><li><code>candidates::AbstractVector{Integer}</code>: Indices of the candidate loci.</li></ul><p><strong>Returns</strong></p><ul><li>A Vector{Float64} of length N with the Geometric genomic offset values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/Geometric_GO.jl#L84-L96">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GeometricGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real" href="#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GeometricGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real"><code>GenomicOffsets.genomic_offset</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">genomic_offset(model::GeometricGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T&lt;:Real</code></pre><p>Compute the Geometric genomic offset.</p><p><strong>Arguments</strong></p><ul><li><code>model::GeometricGO</code>: A GeometricGO model.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.</li><li><code>Xpred::Matrix{T}</code>: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.</li></ul><p><strong>Returns</strong></p><ul><li>A Vector{Float64} of length N with the Geometric genomic offset values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/Geometric_GO.jl#L52-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GradientForestGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real" href="#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GradientForestGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real"><code>GenomicOffsets.genomic_offset</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">genomic_offset(model::GradientForestGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T&lt;:Real</code></pre><p>Compute the Gradient Forest genomic offset. Attention! Our implementation does not perform extrapolation, which means that you may obtain unpredictable results if you try so. </p><p><strong>Arguments</strong></p><ul><li><code>model::GradientForestGO</code>: A GeometricGO model.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.</li><li><code>Xpred::Matrix{T}</code>: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.</li></ul><p><strong>Returns</strong></p><ul><li>A Vector{Float64} of length N with the Gradient forest genomic offset values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/GradientForest_GO.jl#L28-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RDAGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real" href="#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RDAGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real"><code>GenomicOffsets.genomic_offset</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">genomic_offset(model::RDAGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T&lt;:Real</code></pre><p>Compute the RDA genomic offset.</p><p><strong>Arguments</strong></p><ul><li><code>model::RDAGO</code>: A RDAGO model.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li><li><code>Xpred::Matrix{T}</code>: NxP matrix with the predicted / altered environmental matrix.</li><li><code>weighted::Bool=false</code>: If true, the euclidean distance in the projected space is weighted based on the eigenvalue of each axis.</li></ul><p><strong>Returns</strong></p><ul><li>A Vector{Float64} of length N with the RDAGO genomic offset values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RDA_GO.jl#L38-L51">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RONA, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real" href="#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RONA, AbstractMatrix{T}, AbstractMatrix{T}}} where T&lt;:Real"><code>GenomicOffsets.genomic_offset</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">genomic_offset(model::RONA, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T&lt;:Real</code></pre><p>Compute the genomic offset for the RONA model.</p><p><strong>Arguments</strong></p><ul><li><code>model::RONA</code>: A RONA model.</li><li><code>X::AbstractMatrix{T}</code>: NxP matrix with environmental data.</li><li><code>Xpred::Matrix{T}</code>: NxP matrix with the predicted / altered environmental matrix.</li></ul><p><strong>Returns</strong></p><ul><li>A Vector{Float64} of length N with the RONA genomic offset values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/GenomicOffsets.jl/blob/a0d962f1024e2480ee21aa5362faf8b49669d34a/src/RONA.jl#L27-L39">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Friday 24 May 2024 13:59">Friday 24 May 2024</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
