var documenterSearchIndex = {"docs":
[{"location":"#GenomicOffsets.jl-Documentation","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"","category":"section"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"The GenomicOffsets Julia package implements efficient versions of some of the most popular Genomic offset metrics.","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"Disclaimer: this documentation is not finished and just provides a quick overview of the syntax. ","category":"page"},{"location":"#Getting-started","page":"GenomicOffsets.jl Documentation","title":"Getting started","text":"","category":"section"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"The package provides a miminal dataset you can use. You can load it as follos. ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets\nY, X, Xpred = data; ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"On the one hand, Y is the genotype matrix (or allele frequencies) of the current locally adapted individuals (or populations). We have 100 individuals and 500 loci. ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets # hide\nY, X, Xpred = data; # hide\nsize(Y)","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"On the other hand, X is the current environmental matrix. We have 100 individuals and 2 ecological predictors.  Xpred are the same individuals (although this is not mandatory) for which we have a forecasted ecological predictors.  ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets # hide\nY, X, Xpred = data; # hide\n@assert size(X) == size(Xpred)\nsize(X)","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"We can compute the RONA (and the rest of supported methods) by first, fitting the model and then computing the predicted genomic offset. ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets # hide\nY, X, Xpred = data # hide\nrona = fit(RONA, Y, X)\ngenomic_offset(rona, X, Xpred)","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"Other genomic offset metrics have additional parameters. For example, if computing the geometric genomic offset, you may want to provide the number of latent factors of the LFMM model (and not use the Tracy-Widom test to determined it). ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets # hide\nY, X, Xpred = data # hide\ngeometric = fit(GeometricGO, Y, X, 3)\ngenomic_offset(geometric, X, Xpred)","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"To see a more complicated use case, let's compute the 95% confidence niterval using  bootstrapping. We do it by computing the geometric genomic offset repeatly with different loci sampled with replacement. At every iteration, (a) the number of latent factors will be determined using the TracyWidom test, (b) a set of putatively adaptative loci will be identified with an F-test (corrected using genomic control) based on the LFMM model and (c) the geometric genomic offste is computed.  ","category":"page"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"using GenomicOffsets # hide\nY, X, Xpred = data # hide\nusing Random, StatsBase\nnboots = 100\noffsets = bootstrap_with_candidates(GeometricGO, Y, X, Xpred, nboots;candidates_threshold=0.05)\n@assert size(offsets) == (size(Y, 1), nboots) \nconfint95 = [quantile(ind, [0.025, 0.975]) for ind in eachrow(offsets)]","category":"page"},{"location":"#References","page":"GenomicOffsets.jl Documentation","title":"References","text":"","category":"section"},{"location":"","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.jl Documentation","text":"Modules = [GenomicOffsets]\nOrder   = [:function, :type]","category":"page"},{"location":"#GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GenomicOffsets.LFMM, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.LFMM_Ftest","text":"LFMM_Ftest(model::LFMM, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; genomic_control::Bool=true, center=true) where {T1<:Real,T2<:Real}\n\nCompute F-Statistic (F-test) for the Linear Fixed Effects Mixed Model (LFMM). TODO: Add reference & description.\n\nArguments\n\nmodel: A LFMM{T<:Real} data structure (obtained from RidgeLFMM).\nY: A centered genotype matrix of size NxL.\nX: A centered environmental matrix of size NxP.\ngenomic_control: If true, apply genomic control to the F-statistic.\ncenter: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.\n\nReturns\n\nA vector of p-values for loci in the genotype matrix.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.LFMM_Ftest-Union{Tuple{T2}, Tuple{T1}, Tuple{GeometricGO, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.LFMM_Ftest","text":"LFMM_Ftest(model::LFMM, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; genomic_control::Bool=true, center=true) where {T1<:Real,T2<:Real}\n\nCompute F-Statistic (F-test) for the Linear Fixed Effects Mixed Model (LFMM). TODO: Add reference & description.\n\nArguments\n\nmodel: A LFMM{T<:Real} data structure (obtained from RidgeLFMM).\nY: A centered genotype matrix of size NxL.\nX: A centered environmental matrix of size NxP.\ngenomic_control: If true, apply genomic control to the F-statistic.\ncenter: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.\n\nReturns\n\nA vector of p-values for loci in the genotype matrix.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.RidgeLFMM-Union{Tuple{T2}, Tuple{T1}, Tuple{Matrix{T1}, Matrix{T2}, Int64}, Tuple{Matrix{T1}, Matrix{T2}, Int64, Any}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.RidgeLFMM","text":"RidgeLFMM(Y::Matrix{T}, X::Matrix{T}, K::Int, λ::Float64) where T<:Real\n\nRidge solutions for the Linear Fixed Effects Mixed Model (LFMM) as described by Caye et al. (2019).\n\nY = X B^T + W = X B^T + U V^T\n\nArguments\n\nY: A centered genotype matrix of size NxL.\nX: Environmental matrix of size NxP.\nK: Number of latent factors.\nλ: Regularization parameter (by 1e-5)\ncenter: If true, both the genotype matrix and the environmental matrix are centered. If false, both matrices are assumed to be centered.\n\nReturns\n\nA LFMM{T<:Real} data structure with the latent factors U, Vt, and the effect sizes Bt.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.TracyWidom-Union{Tuple{AbstractVector{T}}, Tuple{T}} where T<:Real","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.TracyWidom","text":"TracyWidom(eigenvalues::AbstractVector{T}; sort_eigenvalues::Bool=true) where T<:Real\n\nCompute the Tracy-Widom statistics and p-values for a given set of eigenvalues. TODO: add reference. \n\nArguments\n\neigenvalues::AbstractVector{T}: A vector of eigenvalues. \nsort_eigenvalues::Bool=true: If true, sort the eigenvalues in descending order. If false, the eigenvalues are assumed to be sorted in descending order.\n\nReturns\n\nA tuple with the Tracy-Widom statistics and p-values. For a given threshold, it is advised not to keep all eigenvalues below the threshold, but up to the first one that is above the threshold (not included).\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.bootstrap_with_candidates","text":"bootstrap_with_candidates(::Type{GeometricGO}, rng::Random.AbstractRNG,Y::AbstractMatrix{T1}, X::AbstractMatrix{T2},Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1<:Real,T2<:Real}\n\nCompute the Geometric genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of Y. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.\n\nArguments\n\nrng::Random.AbstractRNG: Random number generator. If not provided, the global RNG is used.\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T2}: NxP matrix with environmental data.\nXpred::AbstractMatrix{T2}: NxP matrix with the predicted / altered environmental matrix.\nnboot::Int=500: Number of bootstrap iterations.\nOptional arguments:\ncandidates_threshold::Real=0.05: Threshold for the F-test. Better to be permisive.\ngenomic_control::Bool=true: Apply genomic control to the F-test.\ntw_threshold::Real=0.001: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).\n\nReturns\n\nA matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{GradientForestGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.bootstrap_with_candidates","text":"bootstrap_with_candidates(::Type{GradientForestGO}, rng::Random.AbstractRNG,Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; ntrees::Int=100, candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1<:Real,T2<:Real}\n\nCompute the Gradient Forest genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of Y. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.\n\nArguments\n\nrng::Random.AbstractRNG: Random number generator. If not provided, the global RNG is used.\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T2}: NxP matrix with environmental data.\nXpred::AbstractMatrix{T2}: NxP matrix with the predicted / altered environmental matrix.\nnboot::Int=500: Number of bootstrap iterations.\nOptional arguments:\nntrees::Int=100: Number of trees each forest (one per allele) will have.\ncandidates_threshold::Real=0.05: Threshold for the F-test. Better to be permisive.\ngenomic_control::Bool=true: Apply genomic control to the F-test.\ntw_threshold::Real=0.001: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).\n\nReturns\n\nA matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RDAGO}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.bootstrap_with_candidates","text":"bootstrap_with_candidates(::Type{RDAGO}, rng::Random.AbstractRNG, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1<:Real, T2<:Real}\n\nCompute the RDA genomic offset using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of Y. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.\n\nArguments\n\nrng::Random.AbstractRNG: Random number generator. If not provided, the global RNG is used.\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T2}: NxP matrix with environmental data.\nXpred::AbstractMatrix{T2}: NxP matrix with the predicted / altered environmental matrix.\nnboot::Int=500: Number of bootstrap iterations.\nOptional arguments:\ncandidates_threshold::Real=0.05: Threshold for the F-test. Better to be permisive.\ngenomic_control::Bool=true: Apply genomic control to the F-test.\ntw_threshold::Real=0.001: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).\n\nReturns\n\nA matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.bootstrap_with_candidates-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}}, Tuple{Type{RONA}, Random.AbstractRNG, AbstractMatrix{T1}, AbstractMatrix{T2}, AbstractMatrix{T2}, Int64}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.bootstrap_with_candidates","text":"bootstrap_with_candidates(::Type{RONA}, rng::Random.AbstractRNG, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}, Xpred::AbstractMatrix{T2}, nboot::Int=500; candidates_threshold::Real=0.05, genomic_control::Bool=true, tw_threshold::Real=0.001) where {T1<:Real, T2<:Real}\n\nCompute the genomic offset for the RONA model using a bootstrap approach. For every, bootstrap iteration, the model is fitted using a random subset of the columns of Y. The Tracy-Widom test is used to estimate the number of latent factors. The F-test is used to select putatively adaptative loci. The genomic offset is computed for the selected loci.\n\nArguments\n\nrng::Random.AbstractRNG: Random number generator. If not provided, the global RNG is used.\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T2}: NxP matrix with environmental data.\nXpred::AbstractMatrix{T2}: NxP matrix with the predicted / altered environmental matrix.\nnboot::Int=500: Number of bootstrap iterations.\nOptional arguments:\ncandidates_threshold::Real=0.05: Threshold for the F-test. Better to be permisive.\ngenomic_control::Bool=true: Apply genomic control to the F-test.\ntw_threshold::Real=0.001: Threshold for the Tracy-Widom test. Better to be conservative (as the number of latent factors is often overconservative).\n\nReturns\n\nA matrix of size NxNboot with the genomic offset values. If no candidate loci are found, the row is filled with zeros.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}, Tuple{Type{GeometricGO}, AbstractMatrix{T1}, AbstractMatrix{T2}, Union{Nothing, Int64}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.fit","text":"fit(::Type{GeometricGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1<:Real,T2<:Real}\n\nFit the Geometric genomic offset model (that is, fit a LFMM) using the data Y and X.\n\nArguments\n\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\nK: Number of latent factors to use when fitting th  e LFMM model. It can be determined using the elbow rule of thumb. If nothing is provided, the Tracy-Widom statistic to determine the number of latent factors (which might be overconservative).\nOptional arguments:\ncenter::Bool=true: Center both the genotype and environmental data. Predicted environmental data is centered using the same mean as X. If false, data is assumed to be centered.\nscale::Bool=true: Scale the environmental data. Predicted environmental data is scaled using the same standard deviation as X. If false, data is assumed to be scaled.\ntw_threshold::Real=1e-3: Threshold to determine the number of latent factors if using the Tracy-Widom statistic. By default, it uses 1e-3.\nλ::Real=1e-5: Regularization parameter for the LFMM.\n\nReturns\n\nA GeometricGO model.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{GradientForestGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.fit","text":"fit(::Type{GradientForestGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}; ntrees::Int=100) where {T1<:Real,T2<:Real}\n\nFit the Gradient forest genomic offset model (that is, fit a Gradient Forest) using the data Y and X.\n\nArguments\n\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\nntrees: Number of trees each forest (one per allele) will have.\n\nReturns\n\nA GradientForestGO model.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RDAGO}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.fit","text":"fit(::Type{RDAGO}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1<:Real,T2<:Real}\n\nFit the RDA model (that is, apply Redundancy Analysis) using the data Y and X.\n\nArguments\n\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\n\nReturns\n\nA RDAGO model.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.fit-Union{Tuple{T2}, Tuple{T1}, Tuple{Type{RONA}, AbstractMatrix{T1}, AbstractMatrix{T2}}} where {T1<:Real, T2<:Real}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.fit","text":"fit(::Type{RONA}, Y::AbstractMatrix{T1}, X::AbstractMatrix{T2}) where {T1<:Real,T2<:Real}\n\nFit the RONA model (that is, solve the least squares problem) using the data Y and X.\n\nArguments\n\nY::AbstractMatrix{T1}: NxL matrix with individuals genotypes or allele frequencies.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\n\nReturns\n\nA RONA model.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.genomic_offset-Union{Tuple{T2}, Tuple{T}, Tuple{GeometricGO{T}, AbstractMatrix{T}, AbstractMatrix{T}, AbstractVector{T2}}} where {T<:Real, T2<:Integer}","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.genomic_offset","text":"genomic_offset(model::GeometricGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T<:Real\n\nCompute the Geometric genomic offset.\n\nArguments\n\nmodel::GeometricGO: A GeometricGO model.\nX::AbstractMatrix{T}: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.\nXpred::Matrix{T}: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.\ncandidates::AbstractVector{Integer}: Indices of the candidate loci.\n\nReturns\n\nA Vector{Float64} of length N with the Geometric genomic offset values.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GeometricGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T<:Real","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.genomic_offset","text":"genomic_offset(model::GeometricGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T<:Real\n\nCompute the Geometric genomic offset.\n\nArguments\n\nmodel::GeometricGO: A GeometricGO model.\nX::AbstractMatrix{T}: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.\nXpred::Matrix{T}: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.\n\nReturns\n\nA Vector{Float64} of length N with the Geometric genomic offset values.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{GradientForestGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T<:Real","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.genomic_offset","text":"genomic_offset(model::GradientForestGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T<:Real\n\nCompute the Gradient Forest genomic offset. Attention! Our implementation does not perform extrapolation, which means that you may obtain unpredictable results if you try so. \n\nArguments\n\nmodel::GradientForestGO: A GeometricGO model.\nX::AbstractMatrix{T}: NxP matrix with environmental data. Data will be scaled and centered if the model was fitted with center and scale set to true.\nXpred::Matrix{T}: NxP matrix with the predicted / altered environmental matrix. Data will be scaled and centered if the model was fitted with center and scale set to true.\n\nReturns\n\nA Vector{Float64} of length N with the Gradient forest genomic offset values.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RDAGO, AbstractMatrix{T}, AbstractMatrix{T}}} where T<:Real","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.genomic_offset","text":"genomic_offset(model::RDAGO, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T<:Real\n\nCompute the RDA genomic offset.\n\nArguments\n\nmodel::RDAGO: A RDAGO model.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\nXpred::Matrix{T}: NxP matrix with the predicted / altered environmental matrix.\n\nReturns\n\nA Vector{Float64} of length N with the RDAGO genomic offset values.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.genomic_offset-Union{Tuple{T}, Tuple{RONA, AbstractMatrix{T}, AbstractMatrix{T}}} where T<:Real","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.genomic_offset","text":"genomic_offset(model::RONA, X::AbstractMatrix{T}, Xpred::AbstractMatrix{T}) where T<:Real\n\nCompute the genomic offset for the RONA model.\n\nArguments\n\nmodel::RONA: A RONA model.\nX::AbstractMatrix{T}: NxP matrix with environmental data.\nXpred::Matrix{T}: NxP matrix with the predicted / altered environmental matrix.\n\nReturns\n\nA Vector{Float64} of length N with the RONA genomic offset values.\n\n\n\n\n\n","category":"method"},{"location":"#GenomicOffsets.GeometricGO","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.GeometricGO","text":"Geometric genomic offset. TODO: Add reference\n\n\n\n\n\n","category":"type"},{"location":"#GenomicOffsets.GradientForestGO","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.GradientForestGO","text":"Gradient genomic offset. TODO: Add reference\n\n\n\n\n\n","category":"type"},{"location":"#GenomicOffsets.LFMM","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.LFMM","text":"LFMM{T<:Real}\n\nA struct to store the results of the Linear Fixed Effects Mixed Model (LFMM) as described by Caye et al. (2019).\n\nY = X B^T + W = X B^T + U V^T\n\n\n\n\n\n","category":"type"},{"location":"#GenomicOffsets.RDAGO","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.RDAGO","text":"Redundancy Analysis (RDA) genomic offset. TODO: Add reference\n\n\n\n\n\n","category":"type"},{"location":"#GenomicOffsets.RONA","page":"GenomicOffsets.jl Documentation","title":"GenomicOffsets.RONA","text":"Risk of non-adaptnesss (RONA) genomic offset. First proposed by Rellstab et al. (2016).\n\n\n\n\n\n","category":"type"}]
}
